### ISSUE TITLE
[Bug]: generateRoute() removing the last asterisk if it's part of the param string

### ISSUE THREAD

**Issue by trainoasis** [remix-run/react-router#9051]
### What version of React Router are you using?

6.3.0

### Steps to Reproduce

```
const path = generatePath("/route/:name", {
    name: "includes *asterisk at the end*",
})
```

### Expected Behavior

I would expect for the above to return `/route/includes *asterisk at the end*`

### Actual Behavior

It returns `/route/includes *asterisk at the end`

**Comment by timdorr** [comment]
I'm not 100% sure this is a valid bug. Yes, it's not doing what you're expecting, but that's also not a valid path, as `*` and spaces aren't valid characters (they should be URL encoded).

**Comment by trainoasis** [comment]
@timdorr thanks for your reply. Actually we used [encodeURIComponent](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/encodeURIComponent) but this [does not URL encode](https://www.w3schools.com/tags/ref_urlencode.asp) `*` by design. 

Should we force URL encode this as a "unique" condition in our code? That also does not seem  very nice to be honest, so I'm a little confused. What do you think?

**Comment by timdorr** [comment]
Sorry, I shouldn't have said asterisks are encoded. They are not according to [the spec](https://www.ietf.org/rfc/rfc3986.txt) (section 2.2). 

You can also use `+` for spaces in URLs. But what you're passing into `generateRoute` should use valid URL characters otherwise. Encoding them is a good first step.

**Comment by trainoasis** [comment]
We are in fact encoding, I just didn't show that in my initial post, we do use `encodeURIComponent` like so: 

```
const path = generatePath("/route/:name", {
    name: encodeURIComponent("includes *asterisk at the end*"),
})
```

Unfortunately the asterisk at the end for some reason gets stripped, so there's only "ugly" solutions such as

1. manually add * it if it's at the end on the first place (after generatePath) 
2. use something like `str.replaceAll("*", "%2A");` before passing to generatePath

I guess it's only me running into this, so it seems will have to take one of the dirty approaches for now. 
If there's any other approaches or actual solutions I can take, I'd be happy to hear them. Thanks!

**Comment by Obi-Dann** [comment]
@trainoasis I've got a PR fixing the issue. `generatePath` will use safer parameter interpolation once it's merged.  https://github.com/remix-run/react-router/pull/10078

**Comment by brophdawg11** [comment]
This was released in [6.9.0](https://github.com/remix-run/react-router/releases/tag/react-router%406.9.0)
